flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -646,6 +646,12 @@ public String toString() {
                 }
 
                 final InputChannel inputChannel = inputChannelOpt.get();
+                if (channelsWithEndOfPartitionEvents.get(
+                        inputChannel.channelInfo.getInputChannelIdx())) {
+                    // channel might have been released by the previous poll:
+                    continue;
+                }
+
```

Comment:
I was thinking if we could also remove the channel from the `channelsWithData` when we receive the EOP. It feels cleaner; however, that would most likely result in additional lock acquisition and I wouldn't be surprised to see some race conditions. Still it would be nice for you to consider.
My second idea is to use `inputChannel.isReleased()` instead to keep it in sync with `queueChannel`. I see that it usually implies a volatile access, so I guess your approach is better. And I'm also assuming that we need to do it in `queueChannel` since this method is called from another thread unlike `waitAndGetNextData`. It still feels suboptimal to check it in two different ways.